### PR TITLE
Adapt cockpit.spec for openSUSE

### PR DESCRIPTION
**Pull request by kkaempf**
This adapt the .spec file to also build on openSUSE
It needs the fix for issue #1780

### CODE REVIEW

**Comment by stefwalter** [comment]
In general we have the cockpit.spec file upstream for use with our CI testing. But I guess, in theory at least, in the future that CI could also happen against multiple OSes. Are you going to maintain the SuSE spec file adaptations upstream consistently?

File: tools/cockpit.spec [hunk 1]
```diff
@@ -71,7 +89,11 @@ BuildRequires: sed
 %endif
 
 # For documentation
+%if 0%{?suse_version}
```
Comment (stefwalter) [inline tools/cockpit.spec]: I think we can just use 'xmlto' on Fedora and RHEL as well.

File: tools/cockpit.spec [hunk 2]
```diff
@@ -167,6 +189,9 @@ cp src/bridge/polkit-workarounds.rules %{buildroot}/%{_datadir}/polkit-1/rules.d
 mkdir -p $RPM_BUILD_ROOT%{_sysconfdir}/pam.d
 install -p -m 644 %{SOURCE1} $RPM_BUILD_ROOT%{_sysconfdir}/pam.d/cockpit
 rm -f %{buildroot}/%{_libdir}/cockpit/*.so
+%if 0%{?suse_version}
+install -d %{buildroot}%{_docdir}/%{name}
```
Comment (stefwalter) [inline tools/cockpit.spec]: Why is this needed? If it's good to do, I think we can do it on Fedora as well.

File: tools/cockpit.spec [hunk 3]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 rm -rf %{buildroot}/%{_datadir}/%{name}/docker
 %endif
+%if 0%{?suse_version}
+%suse_update_desktop_file -u -r -G 'Cockpit Server Manager' %{name} System Utility
+mv %{buildroot}%{_datadir}/doc/%{name} %{buildroot}%{_docdir}/%{name}/html
+%endif
 
 %files
+%if 0%{?suse_version}
+%defattr(-,root,root)
```
Comment (stefwalter) [inline tools/cockpit.spec]: Why isn't this the default? Do all SuSE rpms need this line in each %files section?

File: tools/cockpit.spec [hunk 4]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
+%endif
 %{_docdir}/%{name}/AUTHORS
 %{_docdir}/%{name}/COPYING
 %{_docdir}/%{name}/README.md
+%if 0%{?suse_version}
+%dir %{_datadir}/%{name}
```
Comment (stefwalter) [inline tools/cockpit.spec]: We should probably just do this on Fedora as well.

File: tools/cockpit.spec [hunk 5]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 
 %files bridge
+%if 0%{?suse_version}
+%defattr(-,root,root)
+%endif
 %doc %{_mandir}/man1/cockpit-bridge.1.gz
 %{_bindir}/cockpit-bridge
 %attr(4755, -, -) %{_libexecdir}/cockpit-polkit
+%if 0%{?suse_version}
+%dir %{_libdir}/security
```
Comment (stefwalter) [inline tools/cockpit.spec]: This directory should not be owned by Cockpit. It's the PAM module directory.

File: tools/cockpit.spec [hunk 6]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 %{_datadir}/dbus-1/services/com.redhat.Cockpit.service
 %{_libexecdir}/cockpitd
 
 %files doc
+%if 0%{?suse_version}
+%defattr(-,root,root)
+%endif
+%exclude %{_docdir}/%{name}/AUTHORS
+%exclude %{_docdir}/%{name}/COPYING
+%exclude %{_docdir}/%{name}/README.md
```
Comment (stefwalter) [inline tools/cockpit.spec]: We should probably do this on Fedora as well, no?

File: tools/cockpit.spec [hunk 7]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 %{_datadir}/%{name}/base
 %{_datadir}/%{name}/shell
 %{_datadir}/%{name}/playground
 %{_datadir}/%{name}/server-systemd
 
+%if !0%{?suse_version}
 %post shell
 # HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1185749
 ( cd /var/lib/pcp/pmns && ./Rebuild -du )
+%endif
```
Comment (stefwalter) [inline tools/cockpit.spec]: Are you sure this is not needed on SuSE?

File: tools/cockpit.spec [hunk 8]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 %doc %{_mandir}/man5/cockpit.conf.5.gz
 %doc %{_mandir}/man8/cockpit-ws.8.gz
 %config(noreplace) %{_sysconfdir}/%{name}
 %config(noreplace) %{_sysconfdir}/pam.d/cockpit
 %{_unitdir}/cockpit.service
 %{_unitdir}/cockpit.socket
+%if 0%{?suse_version}
+%dir %{_prefix}/lib/firewalld
+%dir %{_prefix}/lib/firewalld/services
```
Comment (stefwalter) [inline tools/cockpit.spec]: These directories should not be owned by Cockpit. They are firewalld owned directories. If SuSE does not use firewalld, then these files should probably just not be installed.

File: tools/cockpit.spec [hunk 9]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
+%endif
 %{_prefix}/lib/firewalld/services/cockpit.xml
 %{_sbindir}/remotectl
 %{_libexecdir}/cockpit-ws
 %attr(4750, root, cockpit-ws) %{_libexecdir}/cockpit-session
+%if !0%{?suse_version}
```
Comment (stefwalter) [inline tools/cockpit.spec]: It seems like this would break Cockpit's dashboard. The point here is that this directory is writable by the wheel group.

File: tools/cockpit.spec [hunk 10]
```diff
@@ -175,67 +200,121 @@ install -p -m 644 cockpit.pp %{buildroot}%{_datadir}/selinux/targeted/
...
 %pre ws
+%if 0%{?suse_version}
+%service_add_pre %{name}.service
+%endif
 getent group cockpit-ws >/dev/null || groupadd -r cockpit-ws
 getent passwd cockpit-ws >/dev/null || useradd -r -g cockpit-ws -d / -s /sbin/nologin -c "User for cockpit-ws" cockpit-ws
 
 %post ws
+%if 0%{?suse_version}
+%service_add_post %{name}.service
```
Comment (stefwalter) [inline tools/cockpit.spec]: Why does this use the cockpit.service while the Fedora version refers to cockpit.socket? The intent is that cockpit be socket activated. The cockpit.socket is the thing that's enabled and started, and the service not manually enabled or started. At least in the default configuration.

**Comment by stefwalter** [comment]
I'd like to say a big "thank you" for working on SuSE packaging. That's really great. I hope the above comments are helpful.

File: tools/cockpit.spec (hunk 1 above)
Comment (kkaempf) [inline tools/cockpit.spec]: That would be awesome ;-)

File: tools/cockpit.spec (hunk 2 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Well, the doc directory does not get created automatically. Usually "make install DESTDIR=..." should do it,

File: tools/cockpit.spec (hunk 3 above)
Comment (kkaempf) [inline tools/cockpit.spec]: I'd wish this would be the default for SUSE. :-/
Yes, all SUSE rpms need this in each %files section :-(

I need to talk to the release manager (resp. the rpm maintainer) why this isn't the default.

File: tools/cockpit.spec (hunk 6 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Well, SUSE is a bit too strict about file conflicts. rpm allows conflicting files if they have the same size/checksum, SUSE doesn't.

File: tools/cockpit.spec (hunk 7 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Seems to be a pcp-3.10 thing. It's not needed with pcp-3.6.10. I'm still struggling to get pcp-3.10 building on SUSE.

File: tools/cockpit.spec (hunk 8 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Right, these files need to be removed. No firewalld on SUSE.

**Comment by kkaempf** [comment]
Thanks for Cockpit, it's working great on SUSE!

File: tools/cockpit.spec (hunk 10 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Right, my fault. SUSE packaging guidelines only describe .service. I got .socket working meanwhile.

File: tools/cockpit.spec (hunk 2 above)
Comment (stefwalter) [inline tools/cockpit.spec]: Right, but in this case we are assembling that directory manually, see the line below. I don't think it's driven by 'make install'.

File: tools/cockpit.spec (hunk 5 above)
Comment (kkaempf) [inline tools/cockpit.spec]: Right. PAM on SUSE uses /lib64/... instead of /usr/lib64/...

**Comment by stefwalter** [comment]
So once this is done, will it be in a repo, that we can link from http://cockpit-project.org/running.html  ?

**Comment by kkaempf** [comment]
It already is: https://build.opensuse.org/project/show/systemsmanagement:cockpit

**Comment by stefwalter** [comment]
> It already is: https://build.opensuse.org/project/show/systemsmanagement:cockpit

@kkaempf What are the end user instructions for using this build?

**Comment by stefwalter** [comment]
So the reason we have cockpit.spec in our repo at all is for our integration testing. 

So the next question is: "Do we want to host spec files from other distros belong upstream, if they're not being used in CI?"

That said, I favor a pragmatic approach, where while the diff of changes from opensuse -> fedora we can have try a common spec file upstream. However in the future as things get more complex, this may need to change.

**Comment by stefwalter** [comment]
At the end of the day, the reason we have a spec file upstream is for our CI. Only once someone contributes OpenSUSE CI, and wants to use an upstream spec file to drive it ... then we could have the opensuse spec files upstream. Otherwise these files will just bit rot.

I've incorporated several of the (not opensuse specific) changes you've made into the upstream spec file, in order to simplify the diff you have to carry.
